src/libsystemd/sd-device/sd-device.c

```diff
@@ -1475,17 +1475,20 @@ static int device_properties_prepare(sd_device *device) {
         }
 
         if (device->property_tags_outdated) {
-                char *tags = NULL;
+                _cleanup_free_ char *tags = NULL;
+                char *tags_end = NULL;
+                size_t tags_allocated = 0;
                 const char *tag;
 
-                tag = sd_device_get_tag_first(device);
-                if (tag)
-                        tags = strjoina(":", tag);
-
-                while ((tag = sd_device_get_tag_next(device)))
-                        tags = strjoina(tags, ":", tag);
+                if (!greedy_realloc((void**) &tags, &tags_allocated, 2, 1))
+                        return -ENOMEM;
+                tags_end = stpcpy(tags, ":");
 
-                tags = strjoina(tags, ":");
+                for (tag = sd_device_get_tag_first(device); tag; tag = sd_device_get_tag_next(device)) {
+                        if (!greedy_realloc((void**) &tags, &tags_allocated, (tags_end - tags) + strlen(tag) + 1, 1))
```

Comment:
greedy_realloc() is kinda an internal function. The GREEDY_REALLOC() macro (that wraps it) should really be called instead. Is there are reason you caller the fucntion instead of the macro?